src/main/java/com/jcabi/http/wire/AbstractHeaderBasedCachingWire.java

```diff
@@ -0,0 +1,212 @@
+     * @return Response obtained
+     * @throws IOException if fails
+     * @checkstyle ParameterNumber (6 lines)
+     */
+    private Response consultCache(
+        final Request req, final String home, final String method,
+        final Collection<Map.Entry<String, String>> headers,
+        final InputStream content, final int connect, final int read
+    ) throws IOException {
+        final Response rsp;
+        if (this.cache.containsKey(req)) {
+            rsp = this.validateCacheWithServer(
+                req, home, method, headers, content, connect, read
+            );
+        } else {
+            rsp = this.origin.send(
+                req, home, method, headers, content, connect, read
+            );
+            this.updateCache(req, rsp);
+        }
+        return rsp;
+    }
+
+    /**
+     * Check response and update cache if needed.
+     * @todo #90:30min Evict cache entry if If-Modified-Since request responded
+     *  with HTTP_OK code and no Last-Modified header.
+     * @param req Request
+     * @param home URI to fetch
+     * @param method HTTP method
+     * @param headers Headers
+     * @param content HTTP body
+     * @param connect The connect timeout
+     * @param read The read timeout
+     * @return Response obtained
+     * @throws IOException if fails
+     * @checkstyle ParameterNumber (8 lines)
+     */
+    private Response validateCacheWithServer(
+        final Request req, final String home, final String method,
+        final Collection<Map.Entry<String, String>> headers,
+        final InputStream content, final int connect, final int read
+    ) throws IOException {
+        final Response cached = this.cache.get(req);
+        final Collection<Map.Entry<String, String>> hdrs = this.enrich(
+            headers, cached
+        );
+        Response result = this.origin.send(
+            req, home, method, hdrs, content, connect, read
+        );
+        if (result.status() == HttpURLConnection.HTTP_NOT_MODIFIED) {
+            result = cached;
+        } else {
+            this.updateCache(req, result);
+        }
+        return result;
+    }
+
+    /**
+     * Add response to cache.
+     * @param req The request to be used as key
+     * @param rsp The response to add
+     */
+    private void updateCache(final Request req, final Response rsp) {
+        if (rsp.headers().containsKey(this.scvh)) {
+            this.cache.put(req, rsp);
+        }
+    }
+
+    /**
+     * Add identify content version header.
+     *
+     * @param headers Original headers
+     * @param rsp Cached response
+     * @return Map with extra header
+     */
+    private Collection<Map.Entry<String, String>> enrich(
+        final Collection<Map.Entry<String, String>> headers, final Response rsp
+    ) {
+        final List<String> list = rsp.headers().get(
+            this.scvh
+        );
+        final Map<String, String> map =
+            new ConcurrentHashMap<>(headers.size() + 1);
```

Comment:
@gmandnepr this check only makes sense when there's concurrent access, here we don't have it, so PMD could be suppressed. But yes, now I see it was just copied, so let's leave it as it is.